### PR TITLE
Create /register and /logout endpoints. Fixes #11 && second half of #17

### LINKED ISSUE #11: Create /register and /logout endpoints
**Authentication with JWT
For every authentication endpoint, the endpoints should start with `/auth`.
Create and assign this in the `app.js`, and then create the endpoints in the `routes folder.` If the `/auth` endpoint already implemented in the `app.js` then just add the endpoints in the `routes folder`.**
`/signup` is for e-mail registration. While creating a sign up, it should be checked if the same email has been used before. If yes, then they should be given a correct `status code (it can be 403)` and an error message. If not, then the user should be saved successfully and correct `status code (it can be 201)` should be given.

In the `/logout` endpoint, the token should be deleted from the cookie.

NOTE; Unit tests and documentation should be included

### LINKED ISSUE #17: Create authentication validation
Validation for `login` should be conducted according to model schema requirements.
Validation for `sign up` should be conducted according to model schema requirements.

### PR DESCRIPTION
**Pull request by khaldarov**
Endpoint does register to the DB. NOT ready for merge. Requesting review for the user model and server-side validation.

- Added `express-validator` and changed its version.
- Changed user model to make it suitable with 3rd party authentication.

### CODE REVIEW

**Comment by khaldarov** [comment]
Validator does work, but not as intended: it does not show the custom error messages.
Screenshot from 2022-07-27 23-36-05

**Comment by khatibAmjad** [comment]
The username can still be required, I was able to create a username from the email by deleting `@gmail.com` from the email. Other than that the changes you made are enough for the 3rd party authentication.

As for the validator I didn't get the problem, the custom message you wrote is invalid email and that's the message the error gave you on postman. Again maybe I didn't get the question.

One more thing, I think all the server side validation should be done in a middleware not in the controller.

**Comment by khatibAmjad** [comment]
For the user schema we should also remove the `unique: true` from the fields that aren't required, coz i faced an error now telling me that i already have a null phone number ana i can't register another user with the same value

**Comment by khaldarov** [comment]
I applied asked revisions. I also have 2 questions:
- Wouldn't using only the first half of Gmail address as username intervene with existing users? Like I might have a username `adnan`, and when some other user tries to register with the email `adnan@gmail.com`  he will get an error.
- Would removing the `uniqe: true` field be a good approach?
Now I'm thinking of what would've it been like if we had used a different model for 3rd party folks.

Other than that, I didn't use `express-validator`, since it didn't work as I intended.
I will do the validation in a middleware, noted.

**Comment by khaldarov** [comment]
- Finished _/register_ and _/logout_ endpoints; added GET _/register_ render endpoint.
- Added tests for different cases: good path, bad path, and horrible path (lol).
- Fixed server-side validation, making all errors elements of an `errorsArray`.
- Changed user model according to our new needs.
- Added documentation and 3 examples for the _/register_ endpoint; added schema field in documentation.

File: src/controllers/auth.js [hunk 1]
```diff
@@ -1,9 +1,13 @@
 const bcrypt = require('bcrypt');
-// const jwt = require('jsonwebtoken');
 const UserModel = require('../models/user');
 
+const registerPage = (_, res) => {
```
Comment (khatibAmjad) [inline src/controllers/auth.js]: Isn't this frontend responsibility? I don't think we need to render anything.

File: src/models/user.js [hunk 2]
```diff
@@ -135,6 +131,7 @@ const userSchema = new mongoose.Schema(
     },
     avaregeRating: {
       type: Number,
+      required: true,
```
Comment (khatibAmjad) [inline src/models/user.js]: I think this shouldn't be required

File: src/models/user.js [hunk 3]
```diff
@@ -147,16 +144,19 @@ const userSchema = new mongoose.Schema(
     donated: {
       ref: 'Donation',
       type: [objectId],
+      required: true,
```
Comment (khatibAmjad) [inline src/models/user.js]: also the required here will be a problem for the 3rd party authentication

File: src/models/user.js [hunk 4]
```diff
@@ -147,16 +144,19 @@ const userSchema = new mongoose.Schema(
...
       default: [],
     },
     requested: {
       ref: 'Request',
       type: [objectId],
+      required: true,
```
Comment (khatibAmjad) [inline src/models/user.js]: same here

File: src/models/user.js [hunk 5]
```diff
@@ -147,16 +144,19 @@ const userSchema = new mongoose.Schema(
...
       default: [],
     },
     received: {
       ref: 'Request',
       type: [objectId],
+      required: true,
```
Comment (khatibAmjad) [inline src/models/user.js]: same here

File: src/models/user.js [hunk 6]
```diff
@@ -174,14 +174,17 @@ const userSchema = new mongoose.Schema(
     },
     address: {
       type: [addressSchema],
+      required: true,
```
Comment (khatibAmjad) [inline src/models/user.js]: same here

File: src/models/user.js [hunk 7]
```diff
@@ -174,14 +174,17 @@ const userSchema = new mongoose.Schema(
...
       default: [],
     },
     reviews: {
       type: [reviewSchema],
+      required: true,
```
Comment (khatibAmjad) [inline src/models/user.js]: same here

File: src/models/user.js [hunk 8]
```diff
@@ -174,14 +174,17 @@ const userSchema = new mongoose.Schema(
...
       default: [],
     },
     reports: {
       type: [reportSchema],
+      required: true,
```
Comment (khatibAmjad) [inline src/models/user.js]: same here

File: src/docs/doc.yaml [hunk 9]
```diff
@@ -36,19 +30,116 @@ paths:
                   type: string,
                 password1:
                   type: string,
+            examples:
+              'SLIGHTLY BAD PATH':
```
Comment (khatibAmjad) [inline src/docs/doc.yaml]: what's the point of having two different bad entries? aren't they going to give the same result?

File: src/controllers/auth.js [hunk 10]
```diff
@@ -19,39 +23,50 @@ const register = async (req, res) => {
...
-          refugee,
-          provider: 'Local',
-          providerId: 'Local',
-          passwordHash,
-        });
-
-        return res.status(200).redirect('/api/auth/login');
-      }
-      return res.status(401).send({ error: 'Email already exists' });
+    if (usedUsername) {
```
Comment (khatibAmjad) [inline src/controllers/auth.js]: I suggest assigning the username to the user like we did in the 3rd party auth. That way we won't need to check the database twice for existing users.

**Review by khatibAmjad — CHANGES_REQUESTED** [review]
Impressive job! especially with the model and the regex.

File: src/controllers/auth.js (hunk 1 above)
Comment (khaldarov) [reply src/controllers/auth.js]: Sure thing. I can remove that.

File: src/models/user.js (hunk 2 above)
Comment (khaldarov) [reply src/models/user.js]: All the ones that have a `default` will not be a problem when using 3rd party auth. No need to specify them; they'll be added automatically, by default. I wanted to make them look unified. I still can change them if they're causing any issues.

File: src/docs/doc.yaml (hunk 9 above)
Comment (khaldarov) [reply src/docs/doc.yaml]: They return different results. One only checks the format; the other checks the database.

File: src/controllers/auth.js (hunk 10 above)
Comment (khaldarov) [reply src/controllers/auth.js]: I didn't quite get this one. Are you suggesting auto-generated usernames, and allowing users only to choose their email addresses?

Comment (khatibAmjad) [reply src/controllers/auth.js]: exactly

File: src/docs/doc.yaml (hunk 9 above)
Comment (khatibAmjad) [reply src/docs/doc.yaml]: okay

File: src/controllers/auth.js (hunk 10 above)
Comment (khaldarov) [reply src/controllers/auth.js]: Might be a good approach. I can work on that 👍🏽

File: src/models/user.js (hunk 2 above)
Comment (khatibAmjad) [reply src/models/user.js]: I vote for removing the required since they aren't used in this case.

Comment (khaldarov) [reply src/models/user.js]: All the `required: true` fields that also have a `default` are only to make the model look unified; they can be removed. On the other hand, they do not really affect how the database functions.

I thought of the profile photo causing a problem when is removed, but there _will_ be a default, placeholder, image for the profile photo. I still can remove the `required: true` fields if you think they'll cause any issues.

Comment (khaldarov) [reply src/models/user.js]: Removed them 👍🏽

**Comment by khaldarov** [comment]
- Fixed requested revision.
- Added some needed constants.
- Added required documentation.
- Added `sparse` to the model for `phoneNumber` field: it now could be empty.
- Adapted test according to the new conditions.
- Changed default profile photo.

File: src/lib/constants.js [hunk 11]
```diff
@@ -1,4 +1,10 @@
+/* eslint-disable prefer-regex-literals */
 module.exports = Object.freeze({
+  UNIQUE_USERNAME: (email) => `${email.split('@')[0]}_${new Date().valueOf()}`,
```
Comment (khatibAmjad) [inline src/lib/constants.js]: I added this function to a utils fille inside the services folder like this: 
```
const generateUniqeUsername = (email) => {
  return `${email.split('@')[0]}_${new Date().valueOf()}`;
};
```
and then used it in my create method. I think it's better than putting it in the constant file.

**Review by khatibAmjad — CHANGES_REQUESTED** [review]

File: src/__tests__/auth.spec.js [hunk 12]
```diff
@@ -1,5 +1,102 @@
...
+    connectDatabase();
+    await UserModel.findOneAndDelete({ email: 'john@doe.co.uk' });
+  });
+
+  afterAll(async () => {
+    await UserModel.findOneAndDelete({ email: 'john@doe.co.uk' });
+    await mongoose.disconnect();
+  });
+
+  const fakeUser = {
```
Comment (khatibAmjad) [inline src/__tests__/auth.spec.js]: can you change this to `invalidUserInfo`

File: src/__tests__/auth.spec.js [hunk 13]
```diff
@@ -1,5 +1,102 @@
...
+    phoneNumber: '+44 0123456789876543210',
+    age: 21,
+    gender: 'Male',
+    nationality: 'Other',
+    refugee: false,
+    password0: 'qwerty-123',
+    password1: 'Qwerty-123456789',
+  };
+
+  const realUser = {
```
Comment (khatibAmjad) [inline src/__tests__/auth.spec.js]: and this to `validUserInfo`

File: src/controllers/auth.js [hunk 14]
```diff
@@ -1,2 +1,62 @@
...
+      refugee,
+      provider: 'Local',
+      providerId: 'Local',
+      passwordHash,
+    });
+
+    return res.status(201).json({ created: newUser });
+    // return res.status(201).redirect('/api/auth/login');
+  } catch (err) {
+    return res.status(500).send(err);
```
Comment (khatibAmjad) [inline src/controllers/auth.js:52]: I think this should be added to the swagger

File: src/controllers/auth.js [hunk 15]
```diff
@@ -1,2 +1,62 @@
...
+  }
+};
+
+// eslint-disable-next-line arrow-body-style
+const logout = async (req, res) => {
```
Comment (khatibAmjad) [inline src/controllers/auth.js:88]: This also should have a swagger documentation

**Review by khatibAmjad — CHANGES_REQUESTED** [review]

**Review by khatibAmjad — APPROVED** [review]

File: src/controllers/auth.js [hunk 16]
```diff
@@ -1,2 +1,62 @@
...
+    password0,
+  } = req.body;
+
+  try {
+    const usedEmail = await UserModel.findOne({ email });
+    if (usedEmail) {
+      errorsArray.push('Email is already taken');
+    }
+
+    if (errorsArray.length > 0) {
```
Comment (halitbatur) [inline src/controllers/auth.js]: I dont see the usage of this, like before it there's only 1 error that u are checking with and its basicaly returning it as json, so I think this arr can be removed and u can have smth like this
```
if (usedEmail) {
  return res.status(400).json({error:"Email is already taken"})
}
```

File: src/controllers/auth.js [hunk 17]
```diff
@@ -1,2 +1,62 @@
...
+      age,
+      gender,
+      nationality,
+      refugee,
+      provider: 'Local',
+      providerId: 'Local',
+      passwordHash,
+    });
+
+    return res.status(201).json({ created: newUser });
```
Comment (halitbatur) [inline src/controllers/auth.js]: here lets return the newUser right away, no need to nest it insside created key

Comment (halitbatur) [reply src/controllers/auth.js]: plus I dont think its safe to return the newUser with all the data that it has with it

File: src/controllers/auth.js [hunk 18]
```diff
@@ -1,2 +1,62 @@
...
+    // return res.status(201).redirect('/api/auth/login');
```
Comment (halitbatur) [inline src/controllers/auth.js]: no commented code

File: src/lib/constants.js [hunk 19]
```diff
@@ -33,4 +36,33 @@ module.exports = Object.freeze({
...
+  PASSWORD_REGEX: new RegExp(
+    /^(?=.*?[A-Z])(?=(.*[a-z]){1,})(?=(.*[\d]){1,})(?=(.*[\W]){1,})(?!.*\s).{7,}$/
+  ),
+  NAME_REGEX: new RegExp(/^$|\s+/),
+  EMAIL_REGEX: new RegExp(
+    /^[a-zA-Z0-9\-_.]+@[a-z]+\.([a-z]{2,3})+(\.[a-z]{2,3})?$/
+  ),
+  USERNAME_REGEX: new RegExp(/^[a-zA-Z0-9\-_.]{2,20}$/),
+  PHONE_NUMBER_REGEX: new RegExp(/^[+]?[0-9]{1,3}[-\s]?[0-9]{1,3}[0-9]{4,9}$/),
+
```
Comment (halitbatur) [inline src/lib/constants.js]: extra new line

File: src/lib/constants.js [hunk 20]
```diff
@@ -33,4 +36,33 @@ module.exports = Object.freeze({
   ],
   ENUM_POST_TYPE: ['Request', 'Donate'],
   ENUM_POST_STATUS: ['Published', 'Verified'],
+
```
Comment (halitbatur) [inline src/lib/constants.js]: extra new line

File: src/routes/auth.js [hunk 21]
```diff
@@ -1,6 +1,11 @@
 const express = require('express');
+const { registerMiddleware } = require('../middleware/auth');
 const authController = require('../controllers/auth');
 
 const router = express.Router();
 
+router.post('/register', registerMiddleware, authController.register);
+
```
Comment (halitbatur) [inline src/routes/auth.js]: extra new line

**Review by halitbatur — CHANGES_REQUESTED** [review]
Good job Adnan, a few changes and it can be merged

File: src/controllers/auth.js (hunk 16 above)
Comment (khaldarov) [reply src/controllers/auth.js]: I totally forgot about that array. Now fixed.

File: src/lib/constants.js (hunk 19 above)
Comment (khaldarov) [reply src/lib/constants.js:53]: No extra lines allowed. I had `// models' constants end` for that, but removed upon request. Added empty comment lines now instead.

File: src/lib/constants.js (hunk 20 above)
Comment (halitbatur) [reply src/lib/constants.js]: @khaldarov I meant that it was an extra line, so please remove the comment and remove the extra line too

Comment (khaldarov) [reply src/lib/constants.js]: My bad. Now fixed.

**Review by halitbatur — APPROVED** [review]

**Comment by halitbatur** [comment]
@khaldarov Fix conflicts and its ready to be merged

**Pull request merged**